x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/DataTiersUsageTransportAction.java

```diff
@@ -63,72 +67,138 @@ protected void masterOperation(Task task, XPackUsageRequest request, ClusterStat
-                Map<String, List<NodeStats>> tierSpecificNodeStats = separateTiers(nodesStatsResponse);
+                // Determine which tiers each index would prefer to be within
+                Map<String, String> indicesToTiers = tierIndices(indices);
 
-                // Generate tier specific stats for the nodes
-                Map<String, DataTiersFeatureSetUsage.TierSpecificStats> tierSpecificStats = tierSpecificNodeStats.entrySet()
-                    .stream().collect(Collectors.toMap(Map.Entry::getKey, ns -> calculateStats(ns.getValue(), routingNodes)));
+                // Generate tier specific stats for the nodes and indices
+                Map<String, DataTiersFeatureSetUsage.TierSpecificStats> tierSpecificStats = calculateStats(nodesStatsResponse.getNodes(),
+                    indicesToTiers, routingNodes);
 
                 listener.onResponse(new XPackUsageFeatureResponse(new DataTiersFeatureSetUsage(tierSpecificStats)));
             }, listener::onFailure));
     }
 
     // Visible for testing
-    static Map<String, List<NodeStats>> separateTiers(NodesStatsResponse nodesStatsResponse) {
-        Map<String, List<NodeStats>> responses = new HashMap<>();
-        DataTier.ALL_DATA_TIERS.forEach(tier ->
-            responses.put(tier, nodesStatsResponse.getNodes().stream()
-                .filter(stats -> stats.getNode().getRoles().stream()
-                    .map(DiscoveryNodeRole::roleName)
-                    .anyMatch(rn -> rn.equals(tier)))
-                .collect(Collectors.toList())));
-        return responses;
+    // Takes a registry of indices and returns a mapping of index name to which tier it most prefers. Always 1 to 1, some may filter out.
+    static Map<String, String> tierIndices(ImmutableOpenMap<String, IndexMetadata> indices) {
+        Map<String, String> indexByTier = new HashMap<>();
+        indices.forEach(entry -> {
+            String tierPref = entry.value.getSettings().get(DataTierAllocationDecider.INDEX_ROUTING_PREFER);
+            if (Strings.hasText(tierPref)) {
+                String[] tiers = tierPref.split(",");
+                if (tiers.length > 0) {
+                    indexByTier.put(entry.key, tiers[0]);
+                }
+            }
+        });
+        return indexByTier;
     }
 
-    private DataTiersFeatureSetUsage.TierSpecificStats calculateStats(List<NodeStats> nodesStats, RoutingNodes routingNodes) {
+    /**
+     * Accumulator to hold intermediate data tier stats before final calculation.
+     */
+    private static class TierStatsAccumulator {
         int nodeCount = 0;
-        int indexCount = 0;
+        Set<String> indexNames = new HashSet<>();
         int totalShardCount = 0;
         long totalByteCount = 0;
         long docCount = 0;
         final AtomicInteger primaryShardCount = new AtomicInteger(0);
         final AtomicLong primaryByteCount = new AtomicLong(0);
```

Comment:
It's weird to me that we use this as a mutable object where the class values are mutated all over the place, but then for these two we use `Atomic<Thing>`. We can probably switch these to a regular `int` and `long` like the total shard and byte count I think?

Reply:
I thought it strange as well. Will simplify.